torch/_higher_order_ops/cond.py

```diff
@@ -441,16 +420,40 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
     if len(flat_true_outs) != len(flat_false_outs):
         raise RuntimeError("Unmatched number of outputs from cond() branches.")
 
+    merged_outs = []
     for true_out, false_out in zip(flat_true_outs, flat_false_outs):
-        true_meta = _extract_tensor_metadata(true_out)
-        false_meta = _extract_tensor_metadata(false_out)
-        if true_meta != false_meta:
-            raise torch._dynamo.exc.CondOpArgsMismatchError(
-                f"Expected each tensor to have same metadata but got:"
-                f"\n  {true_fn.__name__} returns {true_meta}"
-                f"\n  {false_fn.__name__} returns {false_meta}"
-            )
-    return true_outs
+        merged_outs.append(_merge_tensors(true_out, false_out, mode))
+    return merged_outs
+
+
+def _merge_tensors(a: torch.Tensor, b: torch.Tensor, mode: FakeTensorMode):
+    torch._check(a.dtype == b.dtype)
+    torch._check(a.device == b.device)
+    torch._check(a.dim() == b.dim())
+
+    merged_size = []
+    for s0, s1 in zip(a.size(), b.size()):
+        # TODO: use the symint eq by expr class after rebase
+        if type(s0) is int and type(s1) is int and s0 == s1:
+            merged_size.append(s0)
+        elif (
+            isinstance(s0, torch.SymInt)
+            and isinstance(s1, torch.SymInt)
+            and s0.node.expr == s1.node.expr
+        ):
+            merged_size.append(s0)
+        else:
+            merged_size.append(mode.shape_env.create_unbacked_symint())
+            # TODO: setup value range
+
+    torch._check(a.stride() == b.stride())  # TODO: relax this
+
+    # NYI
+    assert not a.is_quantized
+    assert not b.is_quantized
+
+    with mode:
+        return torch.empty(merged_size, dtype=a.dtype, device=a.device)
```

Comment:
Need to handle strides correclty

Reply:
The striding situation is kind of interesting.

Suppose that we have a return contiguous (2, 5) on true branch and (2, 9) on false branch. Size merging will give us (2, u0). But what do we get for the strides? If they're contiguous, the strides are (5, 1) and (9, 1). We specifically want to *reuse* u0 in the strides, so we have (u0, 1). Et cetera.

The flow here is fairly similar to stride symbol allocation in create_symbolic_sizes_and_strides in ShapeEnv. Need to do something similar here.